### PR TITLE
Block implicit DataVector conversion to Variables

### PR DESCRIPTION
**Pull request by wthrowe**
This would have caught a bug in some code I'm working on.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Review by nilsdeppe — DISMISSED** [review]
LGTM, should we also block `ComplexDataVector`? I don't remember if we have Variables of modal coefficients. If so, probably also the `ModalVector` and `ComplexModalvector`?

No need for me to re-approve anything to merge :)

**Comment by wthrowe** [comment]
Good idea.  I'll check if a blanket `VectorImpl` works.

File: src/DataStructures/Variables.hpp [hunk 1]
```diff
@@ -268,6 +270,13 @@ class Variables<tmpl::list<Tags...>> {
   template <typename VT, bool VF>
   Variables& operator=(const blaze::DenseVector<VT, VF>& expression);
 
+  // Prevent the previous two declarations from implicitly converting
```
Comment (kidder) [inline src/DataStructures/Variables.hpp:273]: would a static_assert using is_convertible have caught this in a test?  I.e. would `static_assert(std::is_convertible<DataVector, Variables<tmpl::list<something>>>)` succeed before and fail now?

Also, how about adding an ASSERT to the converting constructors that test that `(*expression).size() % number_of_independent_components == 0)`

Comment (wthrowe) [reply src/DataStructures/Variables.hpp:273]: The static assertion would have caught it.  Added that and a new commit for the size check.

**Review by kidder — APPROVED** [review]

**Pull request merged**